components/builder-worker/src/runner/job_streamer.rs

```diff
@@ -0,0 +1,419 @@
+        // would be where we tag a line as being from a `stderr` source.
+        self.target
+            .lock()
+            .expect("Stream target mutex is poisoned!")
+            .stream_line(self.id, line)
+    }
+
+    /// Finishes a log streamer by writing any remaining messages, marking the log as complete,
+    /// etc. This method can be called multiple times but will only take action once.
+    ///
+    /// # Panics
+    ///
+    /// * If the stream target mutex is poisoned
+    ///
+    /// # Errors
+    ///
+    /// * If the stream target could not be written to
+    pub fn finish(&mut self) -> Result<()> {
+        // Early return if the section has ended to make sure that the `Drop` implementation
+        // doesn't double-finish the stream.
+        if self.finished {
+            return Ok(());
+        }
+
+        let mut target = self.target.lock().expect(
+            "Stream target mutex is poisoned!",
+        );
+        target.stream_line(
+            self.id,
+            format!("builder_log::end::{}", self.id),
+        )?;
+        self.finished = true;
+        target.finish(self.id)
+    }
+}
+
+impl Drop for JobStreamer {
+    fn drop(&mut self) {
+        // This unwrap is intentional as more error handling isn't possible in a `Drop` trait
+        self.finish().unwrap();
+    }
+}
+
+/// The target to which a log stream is written. This struct wraps a remote socket which is log
+/// line aware.
+struct StreamTarget {
+    /// A zeromq socket which represents the log stream target
+    pub sock: zmq::Socket,
+    /// The current line count of submitted log lines
+    pub line_count: u64,
+    /// A local file logger that writes a copy of each line written to the remote socket
+    pub local_logger: Logger,
+}
+
+impl StreamTarget {
+    /// Constructs a new stream target with an initialized socket.
+    ///
+    /// # Panics
+    ///
+    /// * If the zeromq socket cannot be fully set up
+    fn new(workspace: &Workspace) -> Self {
+        let sock = (**DEFAULT_CONTEXT).as_mut().socket(zmq::PUSH).unwrap();
+        sock.set_immediate(true).unwrap();
+        sock.set_linger(5000).unwrap();
+        sock.connect(INPROC_ADDR).unwrap();
+
+        let id = workspace.job.get_id().to_string();
+        let mut local_logger = Logger::init(workspace.root(), format!("local-stream-{}.log", &id));
+        local_logger.log_ident(&id);
+
+        StreamTarget {
+            sock,
+            line_count: 0,
+            local_logger,
+        }
+    }
+
+    /// Takes a string, interpreted as a single line, with a job identifier and writes it to the
+    /// log stream on the socket.
+    ///
+    /// # Panics
+    ///
+    /// * If the protobuf struct cannot be serialized into bytes
+    ///
+    /// # Errors
+    ///
+    /// * If a message couldn't be sent successfully to the stream target socket
+    fn stream_line<S: Into<String>>(&mut self, id: u64, line: S) -> Result<()> {
```

Comment:
Alternatively, could this be
```rust
fn stream_line(&mut self, id: u64, line: &str) -> Result<()> {
    …
    let mut line: String = line.into();
    // or
    let mut line = line.to_string();
```
?

Eliminating the `<S: Into<String>>` makes things a bit simpler, but I wasn't sure if that provided something I didn't understand.